packages/teleport-uidl-validator/src/validator/utils.ts

```diff
@@ -99,6 +99,24 @@ export const checkDynamicDefinitions = (input: Record<string, unknown>) => {
   const errors: string[] = []
 
   UIDLUtils.traverseNodes(input.node as UIDLNode, (node) => {
+    if (node.type === 'element' && node.content.events) {
```

Comment:
Looks good 👍, but please don't use // @ts-ignore in the code. Tests it's fine, since we are importing `.json` files. And the checks and error message can be improved a little.

```suggestion
    if (node.type === 'element' && node.content.events) {
      Object.keys(node.content.events).forEach((eventKey) => {
        node.content.events[eventKey].forEach((event) => {
          if (!stateKeys.includes(event.modifies)) {
            const errorMsg = `"${event.modifies}" is used in events, but not defined`
            errors.push(errorMsg)
            return
          }
      })
    }
```

Reply:
Let's remove the type-check. Sorry for that. Here is the reason, we have two type of handlers defined in event. One is `toggle` type that modifies a state. And second is an event handler. So, for now, let's remove the type check

```js
{
  "click": [
    {
      "type": "propCall",
      "calls": "onClose"
    },
    {
      "type": "stateChange",
      "modifies": "fakeState",
      "newState": false
    }
  ]
}
```

As you can see, the handlers of two types. `stateChange` and `propCall`. The check for `modifies` should happen when it is `stateChange` and the `newState` is not equal to `$toggle`

If it is `$toggle` the state that modifies should be of `boolean`. Since, we can toggle only a boolean.
```js
{
  "click": [
    {
      "type": "stateChange",
      "modifies": "isExpanded",
      "newState": "$toggle"
    }
  ]
}
```

And one more last thing, can you try extending the `decoders` to check for events too. But that's optional 😄. I can take care if you want.  But here is where you can take a look
https://github.com/teleporthq/teleport-code-generators/blob/development/packages/teleport-uidl-validator/src/decoders/utils.ts#L287
You can define `propCall` and `stateChange` events and make a union out of it. 

The events area of the project is not documented so great. So, it will be a good start and great help with the PR. Thank you 😄

Feel free to ping anytime for any doubts, and you can run `yarn test:watch` or `yarn test` to cross-check the test-cases 😄

Reply:
I'll be happy to do that @JayaKrishnaNamburu, I think union events will be defined as
`
export const eventHandlerStatementDecoder: Decoder<UIDLEventHandlerStatement> = object({
  type: union(constant("stateChange"), constant('propCall')),
  calls: optional(string()),
  modifies: optional(string()),
  newState: optional(union(string(), number(), boolean())),

  args: optional(array(union(string(), number(), boolean()))),
})
`
right?

Reply:
This works, but not intuitive at first glance or working with it. It will be more clear if we approach this something like

in teleport-types/uidl.ts

```js
export type UIDLEventHandlerStatement = UIDLPropCallEvent | UIDLStateModifierEvent
```

```js
export interface UIDLPropCallEvent {
  type: 'propCall'
  calls?: string
  args?: Array<string | number | boolean>
}
export interface UIDLStateModifierEvent {
  type: 'stateChange'
  modifies?: string
  newState?: string | number | boolean
}

export type UIDLEventHandlerStatement = UIDLPropCallEvent | UIDLStateModifierEvent
```

And then extend on the `validators` and the same for the validations inside the `validators` in `validateComponentSchema` too.